python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.types.v2;
+
+import java.util.ArrayDeque;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.function.Consumer;
+import org.sonar.plugins.python.api.LocationInFile;
+import org.sonar.python.semantic.v2.SymbolsModuleTypeProvider;
+
+public class LazyType implements PythonType {
+
+  String fullyQualifiedName;
+  private final Queue<Consumer<PythonType>> consumers;
+  private final SymbolsModuleTypeProvider symbolsModuleTypeProvider;
+
+  public LazyType(String fullyQualifiedName, SymbolsModuleTypeProvider symbolsModuleTypeProvider) {
+    this.fullyQualifiedName = fullyQualifiedName;
+    this.symbolsModuleTypeProvider = symbolsModuleTypeProvider;
+    consumers = new ArrayDeque<>();
+  }
+
+  public String fullyQualifiedName() {
+    return fullyQualifiedName;
+  }
+
+  public LazyType addConsumer(Consumer<PythonType> consumer) {
+    consumers.add(consumer);
+    return this;
+  }
+
+  public LazyType resolve(PythonType type) {
+    consumers.forEach(c -> c.accept(type));
+    consumers.clear();
+    return this;
+  }
+
+  public PythonType resolve() {
+    PythonType resolvedType = symbolsModuleTypeProvider.resolveLazyType(this);
+    consumers.forEach(c -> c.accept(resolvedType));
+    consumers.clear();
+    return resolvedType;
+  }
+
+  @Override
+  public PythonType unwrappedType() {
+    return this.resolve().unwrappedType();
```

Comment:
since the current idea is that we never operate with lazy type outside of TypeInference, possibly we don't need to override this methods or need to throw an exception here